### PR TITLE
network: Add Capabilities to Kademlia database

### PR DESCRIPTION
**Pull request by nolash**
This PR makes the Capabilities bitvector of each peer record stored in kademlia available to client code. It introduces the possibility of adding separate indices as pots for the kademlia, which will be added to and removed with calls to `On` and `Off` if the corresponding capability filter matches the entry being manipulated.

The capabilities are embedded in the `BzzAddr` object, which makes it convenient to transmit them through the kademlia aswell as the handshake message.

It also contains the following changes:

* Omit self-connect in `cmd/swarm/run_test.go`
* Add constructor for `BzzAddr` to avoid null capabilities.
* Explicit rename of `BzzAddr` related methods to include "Bzz" in name
* Custom RLP deserialization of `peerMsg` and `BzzAddr` due to failure of default operation to handle the capabilities array.
* Enforce deterministic `BzzAddr` and `enode.ID` relation in test peer generation in `network`.
* Remove some moot vars in `Kademlia` methods.
* Remove obsolete lightnode test
* Add comments to previously uncommented code from ENR/enode refactor and Capabilities implementation

Protocol versions are bumped to `bzz:13` (since nerds after all greet any superstitions with contempt) and `hive:11`

### CODE REVIEW

**Review by zelig — APPROVED** [review]

File: network/kademlia.go [hunk 1]
```diff
@@ -109,11 +112,92 @@ func NewKademlia(addr []byte, params *KadParams) *Kademlia {
...
+	ePeer, ok = p.(*Peer)
+	if ok {
+		eAddr = ePeer.BzzAddr
+	} else if disconnectOnly {
+		return
+	} else {
+		eAddr = p.(*entry).BzzAddr
+	}
+	for s, idxItem := range k.capabilityIndex {
+		if ok {
```
Comment (kortatu) [inline network/kademlia.go:185]: The code in these two ifs seems almost the same. Could not be it improved to be more compact?
Maybe a new function` idxItem.swap(p, s)`?

```go
func (ci *capabilityIndex) swap(p interface{}, s string) {
	var index *pot.Pot
	switch p.(type) {
	case *Peer:
		index = ci.conns
	case *BzzAddr:
		index = ci.addrs
	}
	newIndx, _, found, _ := pot.Swap(index, p, Pof, func(_ pot.Val) pot.Val {
		return nil
	})
	if found {
		log.Trace("Removed peer from capability index", "s", s, "p", p)
		switch p.(type) {
		case *Peer:
			ci.conns = newIndx
		case *entry:
			ci.addrs = newIndx
		}
	}
}
```

So the code in the loop would be substituted by
```go
	for s, idxItem := range k.capabilityIndex {
		if ok {
			idxItem.swap(ePeer, s)
		}
		if !disconnectOnly {
			idxItem.swap(eAddr, s)
		}
```

File: network/kademlia.go [hunk 2]
```diff
@@ -833,7 +957,7 @@ func (k *Kademlia) knowNeighbours(addrs [][]byte) (got bool, n int, missing [][]
 	pm := make(map[string]bool)
 	depth := depthForPot(k.conns, k.NeighbourhoodSize, k.base)
 	// create a map with all peers at depth and deeper known in the kademlia
-	k.eachAddr(nil, 255, func(p *BzzAddr, po int) bool {
+	k.eachAddr(nil, nil, 255, func(p *BzzAddr, po int) bool {
```
Comment (kortatu) [inline network/kademlia.go]: Why not `k.EachAddr(nil, 255, func(p *BzzAddr, po int) bool {`, and avoid that misterious `nil`?
Or even `k.eachAddr(nil, k.addrs, 255, func(p *BzzAddr, po int) bool {` seems more appropiate.

File: network/kademlia.go [hunk 3]
```diff
@@ -872,7 +996,7 @@ func (k *Kademlia) connectedNeighbours(peers [][]byte) (got bool, n int, missing
 	// in order deepest to shallowest compared to the kademlia base address
 	// all bins (except self) are included (0 <= bin <= 255)
 	depth := depthForPot(k.conns, k.NeighbourhoodSize, k.base)
-	k.eachConn(nil, 255, func(p *Peer, po int) bool {
+	k.eachConn(nil, nil, 255, func(p *Peer, po int) bool {
```
Comment (kortatu) [inline network/kademlia.go:1002]: Sames as above, maybe replace with `EachConn`

File: network/kademlia_test.go [hunk 4]
```diff
@@ -670,3 +643,261 @@ func TestKademlia_SubscribeToNeighbourhoodDepthChange(t *testing.T) {
 		}
 	})
 }
+
+// TestCapabilitiesIndex checks that capability indices contains only the peers that have the filters' capability bits set
+func TestCapabilityIndex(t *testing.T) {
```
Comment (kortatu) [inline network/kademlia_test.go:655]: Maybe this test method could split in several small testCases functions specifying what is been under test in each one. It is hard to follow all test cases as the function  advances.

File: network/protocol.go [hunk 5]
```diff
@@ -265,7 +264,7 @@ func (b *Bzz) RunProtocol(spec *protocols.Spec, run func(*BzzPeer) error) func(*
 			Peer:       protocols.NewPeer(p, rw, spec),
 			BzzAddr:    handshake.peerAddr,
 			lastActive: time.Now(),
-			LightNode:  isLightCapability(handshake.Capabilities.Get(0)), // this is a temporary member kept until kademlia code accommodates Capabilities instead
+			LightNode:  isLightCapability(handshake.peerAddr.Capabilities.Get(0)), // this is a temporary member kept until kademlia code accommodates Capabilities instead
```
Comment (kortatu) [inline network/protocol.go]: Is this still needed?
According to the comment, it shouldn't be needed anymore.
Instead of checking this boolean we should be checking BzzAddr.capabilities, right?

**Review by kortatu — COMMENTED** [review]
LGTM, except for the LightNode field in BzzPeer, that I think we don't need anymore.

File: network/kademlia.go (hunk 3 above)
Comment (nolash) [reply network/kademlia.go:1002]: Less cryptic is better.

We use lowercase method because the calling code is responsible for the lock.

File: network/kademlia.go (hunk 1 above)
Comment (nolash) [reply network/kademlia.go:185]: We could, but that's indirection per index instead of just once before the loop. This code does not need to be the most performant, perhaps, but I'm also not sure how much more clarity the change brings. What do you think @zelig ?

File: network/kademlia.go [hunk 6]
```diff
@@ -109,11 +112,92 @@ func NewKademlia(addr []byte, params *KadParams) *Kademlia {
...
+	} else if _, ok := k.capabilityIndex[s]; ok {
+		return fmt.Errorf("Capability index '%s' already exists", s)
+	}
+	log.Debug("Registered cap index", "s", s, "c", c)
+	k.capabilityIndex[s] = NewCapabilityIndex(c)
+	return nil
+}
+
+// adds a peer to any capability indices it matches
+func (k *Kademlia) addToCapabilityIndex(p interface{}) {
```
[the remark this answers is not in the document]
Comment (nolash) [reply network/kademlia.go:144]: I think this is a good idea. But that makes me wonder why the record types in the kademlia are two different ones in the first place, and not just merely `entry` in all cases. Let's make this a separate refactor pass.

File: network/kademlia_test.go (hunk 4 above)
Comment (nolash) [reply network/kademlia_test.go:655]: @kortatu I've split up the test. I am not totally sure how much clearer the split up makes the narrative. But at least I think the cap string mapping on the peers was an improvement.

File: network/kademlia.go (hunk 1 above)
Comment (kortatu) [reply network/kademlia.go:185]: In fact that swap method (although I should call it remove) could be split in two methods removeAddress and removeConnection, so no need of switch at the start of the method is needed: 
```go
func (ci *capabilityIndex) removeAddress(addr *BzzAddr, s string) {
	newIndx, _, found, _ := pot.Swap(ci.addrs, addr, Pof, func(_ pot.Val) pot.Val {
		return nil
	})
	if found {
		ci.addrs = newIndx
	}
}

func (ci *capabilityIndex) removeConnection(conn *Peer, s string) {
	newIndx, _, found, _ := pot.Swap(ci.conns, conn, Pof, func(_ pot.Val) pot.Val {
		return nil
	})
	if found {
		ci.conns = newIndx
	}
}


```

Comment (nolash) [reply network/kademlia.go:185]: That is true. But I'm still not convinced it contributes to increased clarity, as now you have the two function calls separately anyway :) I would prefer to leave it as it is, if you don't mind.

Comment (kortatu) [reply network/kademlia.go:185]: Ok, I think the problem is more like you said that we are using two different types in the indexes instead of always using *entry.

**Review by kortatu — APPROVED** [review]

File: network/discovery.go [hunk 7]
```diff
@@ -127,6 +130,27 @@ type peersMsg struct {
 	Peers []*BzzAddr
 }
 
+// DecodeRLP implements rlp.Decoder interface
```
Comment (zelig) [inline network/discovery.go:133]: why is this needed?

File: network/network.go [hunk 8]
```diff
@@ -3,17 +3,76 @@ package network
...
 // BzzAddr implements the PeerAddr interface
 type BzzAddr struct {
-	OAddr []byte
-	UAddr []byte
+	OAddr        []byte
+	UAddr        []byte
+	Capabilities *capability.Capabilities
+}
+
+// EncodeRLP implements rlp.Encoder
```
Comment (zelig) [inline network/network.go:23]: it should be enough to implement custom rlp for Capabilities within that package no?

File: network/network.go [hunk 9]
```diff
@@ -3,17 +3,76 @@ package network
...
+		return err
+	}
+	err = rlp.Encode(w, y)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// EncodeRLP implements rlp.Decoder
```
Comment (zelig) [inline network/network.go]: Encode -> Decode if it stays

File: network/network_test.go [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+	b, err := rlp.EncodeToBytes(addr)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var addrRecovered BzzAddr
+	err = rlp.DecodeBytes(b, &addrRecovered)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
```
Comment (zelig) [inline network/network_test.go:42]: maybe compare with the orig to complete the test

**Review by zelig — APPROVED** [review]
can we get rid of custom serialisations?

File: network/network.go (hunk 8 above)
Comment (nolash) [reply network/network.go:23]: I thought so too, but it's not

**Comment by nolash** [comment]
@zelig I intend to make another pass later on the `capability` package to make the bitflag for the wire directly available behind a method call. I wish I left it the way it was originally, really...
